### PR TITLE
Add an extensible API for implicit casts

### PR DESCRIPTION
**Pull request by pitrou**
(as proposed in NBEP 2)
Based on PR #1575.

### CODE REVIEW

**Comment by seibert** [comment]
#1575 is merged to master now, so updating this PR should shorten the diff.

**Comment by seibert** [comment]
Looks good to me.  Any comments, @sklam?

File: numba/targets/builtins.py [hunk 1]
```diff
@@ -1324,6 +1312,96 @@ def complex_impl(context, builder, sig, args):
...
+    """
+    Convert any number to any other.
+    """
+    [val] = args
+    [valty] = sig.args
+    return context.cast(builder, val, valty, sig.return_type)
+
+
+#-------------------------------------------------------------------------------
+# Implicit casts between numerics
```
Comment (stuartarchibald) [inline numba/targets/builtins.py]: Are these implicit cast combinations exhaustively tested somewhere?

Comment (pitrou) [inline numba/targets/builtins.py]: Le 30/12/2015 19:43, stuartarchibald a écrit :
> 
> Are these implicit cast combinations exhaustively tested somewhere?

There's test_numconv and test_numberctor, yes.

**Comment by sklam** [comment]
Looks good

**Comment by seibert** [comment]
With the massive merge from the JIT classes PR, this PR now conflicts with master.  Once you resolve the conflict, I'll merge this as well.

**Comment by pitrou** [comment]
Merge done.

**Pull request merged**